discord/app_commands/commands.py

```diff
@@ -129,15 +129,67 @@
 VALID_SLASH_COMMAND_NAME = re.compile(r'^[\w-]{1,32}$')
 CAMEL_CASE_REGEX = re.compile(r'(?<!^)(?=[A-Z])')
 
+ARG_NAME_SUBREGEX = r"(?:\\?\*)*(?P<name>[^\s:\-]+)"
+
+ARG_DESCRIPTION_SUBREGEX = r"(?P<description>(?:.|\n)+?(?:\Z|\r?\n(?=[\S\r\n])))"
+
+ARG_TYPE_SUBREGEX = r"(?P<type>.+)"
+
+GOOGLE_DOCSTRING_ARG_REGEX = re.compile(
+    rf"^{ARG_NAME_SUBREGEX}[ \t]*(?:\({ARG_TYPE_SUBREGEX}\))?[ \t]*:[ \t]*{ARG_DESCRIPTION_SUBREGEX}",
+    re.MULTILINE,
+)
+
+SPHINX_DOCSTRING_ARG_REGEX = re.compile(
+    rf"^:param {ARG_NAME_SUBREGEX}:[ \t]+{ARG_DESCRIPTION_SUBREGEX}[ \t]*(?::type [^\s:]+:[ \t]+{ARG_TYPE_SUBREGEX})?",
+    re.MULTILINE,
+)
+
+NUMPY_DOCSTRING_ARG_REGEX = re.compile(
+    rf"^{ARG_NAME_SUBREGEX}(?:[ \t]*:)?(?:[ \t]+{ARG_TYPE_SUBREGEX})?[ \t]*\r?\n[ \t]+{ARG_DESCRIPTION_SUBREGEX}",
+    re.MULTILINE,
+)
+
 
 def _shorten(
     input: str,
     *,
     _wrapper: TextWrapper = TextWrapper(width=100, max_lines=1, replace_whitespace=True, placeholder='…'),
 ) -> str:
+    try:
+        # split on the first double newline since arguments may appear after that
+        input, _ = re.split(r'\n\s*\n', input, maxsplit=1)
+    except ValueError:
+        pass
     return _wrapper.fill(' '.join(input.strip().split()))
 
 
+def _parse_args_from_docstring(func: Callable[..., Any], params: Dict[str, CommandParameter]) -> Dict[str, str]:
+    docstring = inspect.getdoc(func)
+
+    if docstring is None:
+        return {}
+
+    # Extract the arguments
+    # Note: These are loose regexes, but they are good enough for our purposes
+    # For Google-style, look only at the lines that are indented
+    section_lines = inspect.cleandoc("\n".join(line for line in docstring.splitlines() if line.startswith("  ")))
+    docstring_styles = (
+        GOOGLE_DOCSTRING_ARG_REGEX.finditer(section_lines),
+        SPHINX_DOCSTRING_ARG_REGEX.finditer(docstring),
+        NUMPY_DOCSTRING_ARG_REGEX.finditer(docstring),
+    )
+
+    # Choose the style with the largest number of arguments matched
+    matched_args = []
+    for matches in docstring_styles:
+        style_matched_args = [match for match in matches if match.group("name") in params]
+        if len(style_matched_args) > len(matched_args):
+            matched_args = style_matched_args
+
+    return {match.group("name"): match.group("description") for match in matched_args}
```

Comment:
I wonder if it would be better to do this in a single dict-comprehension instead that way we can leverage the C opcodes. The behaviour is admittedly subtly different but I'm not sure what degenerate case you were thinking of.

```suggestion
    return {
        m.group('name'): m.group('description') for matches in docstring_styles for m in matches if m.group('name') in params
    }
```

Reply:
I think it's a bit less readable, but it looks like it should be fine.

It would allow mixing different styles of docstrings together which is not really a problem since it seems to be working on all the normal docstrings I used for testing.